### PR TITLE
Fix Enabled incorrect draw state for DateTimePicker

### LINKED ISSUE #3533: Disabled DateTimePicker at Design-Time does not have the correct border color when enabled when running
* .NET Core Version: 3.1.400-preview-015178
* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**

When using a DateTimePicker which is initialized as `Enabled = false` in the designer, enabling the control later when the app is running does not change the border color, which stays with a "disabled gray" color.

bug

Note: if the control is enabled by default, and you disable/enable it, the border is changed correctly.

**Expected behavior:**

When enabling the control, the border should change to use the standard border color instead of keeping the disabled one.

**Minimal repro:**

- Create a new Form
- Drop a `DateTimePicker` control
- Set `Enabled` to `false`
- Add a button which change the `Enabled` state to `true`
- Start the project
- Click the button to enable the control.

## Edit based on following comments

**There seems to be a bug in the way the border color of the control is set and drawn by the .net control or the native control.**

* The border color which will be used for the entire existence of the control seems to be set when the handle is created.
* If the control is in an enabled state, a dark gray color will be used. If it is disabled a light gray color will be used instead.
* Once the handle is created, the border color will never change again, despite changing the Enabled property (or sending WM_ENABLE messages)
* The only way to update the border color of the control is to force the recreation of the handle, which will change the border color according to the control state.
* (Not necessary related: there is a strange extended style WS_EX.CLIENTEDGE set on the control, which seems to be useless and break `UpdateStyles()` calls)

### PR DESCRIPTION
**Pull request by tbolon**
Fixes #3533

## Proposed changes

- override OnEnabledChanged() to force style update.

## Customer Impact

- minimal: style update is a lightweight operation

## Regression? 

- No

## Risk

- (TODO)

### Before

When created enabled, the border's color is a dark gray and never changes.
enabled-bug

When created disabled, the border's color is a light gray and never changes.
disabled-bug

### After

Now the border color changes according to the state of the control.

Tested with both initial state enabled and disabled.

enabled-fixed

disabled-fixed

## Test methodology 

- manual testing

## Test environment(s) 

- TODO

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4374)

### CODE REVIEW

File: src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DateTimePicker.cs [hunk 1]
```diff
@@ -17,6 +17,18 @@ public partial class DateTimePicker : Form
...
+            checkBox.Checked = true;
+            checkBox.Location = new Point(27, 60);
+            checkBox.Click += CheckBox_Click;
+            checkBox.Text = "Enabled";
+            this.Controls.Add(checkBox);
+        }
+
+        private void CheckBox_Click(object sender, EventArgs e)
+        {
+            dateTimePicker1.Enabled = !dateTimePicker1.Enabled;
```
Comment (tbolon) [inline src/System.Windows.Forms/tests/IntegrationTests/WinformsControlsTest/DateTimePicker.cs:31]: This change will be removed in a future commit.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        [WinFormsFact]
+        public unsafe void DateTimePicker_VisualStyles_Enabled_Border()
+        {
+            if (!Application.RenderWithVisualStyles)
+            {
+                return;
+            }
+
+            using Form form = new Form();
+            using DateTimePicker control = new DateTimePicker { };
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs:47]:
```suggestion
            using DateTimePicker control = new DateTimePicker();
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Drawing;
+using System.Windows.Forms.Metafiles;
+using Xunit;
+using static Interop;
+
+namespace System.Windows.Forms.Tests
+{
+    public class DateTimePickerRenderingTests : IClassFixture<ThreadExceptionFixture>
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs:12]: For rendering tests we need to have 2 tests: with and without visual styles enabled.

Visual styles rendering goes through the theming API, and we haven't enhanced the test infra to validate that the final result meets expectations. Hence the existing rendering tests with visual styles are currently in non-working state.

To run visual styles disabled scenarios locally you need to comment out `Application.EnableVisualStyles();` in `ThreadExceptionFixture`.
And then you can verify the border is drawn with the expected color via `emf.Validate(...)`.

Holler if you get stuck

File: src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs [hunk 4]
```diff
@@ -1181,6 +1181,17 @@ protected override void OnHandleDestroyed(EventArgs e)
...
         }
 
+        /// <inheritdoc />
+        protected override void OnEnabledChanged(EventArgs e)
+        {
+            base.OnEnabledChanged(e);
+
+            if (IsHandleCreated)
+            {
+                RecreateHandle();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1191]: May be a good idea to add a comment explaining why we're doing this.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs (hunk 3 above)
Comment (tbolon) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs:12]: I have created the tests to render the control with visual styles on or off. The problem with the datetimepicker is that without visual styles enabled, the border is rendered the same way in enabled and disabled state (maybe it explains the problem with visual styles enabled).

So my rendering tests are only valid with Visual Styles enabled.

So should I skip rendering tests for now and only provide screenshots on this PR ?

Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerRenderingTests.cs:12]: Screenshots will certainly help. But I don't think should skip tests, better to add them and decorate with a "Skip" attribute, so we know to come to them.

**Comment by RussKie** [comment]
@JeremyKuhne I've run rendering tests with visual styles off, and the only EMF difference between "enabled" and "disabled" controls is the colour of the text. However visually there are more differences - including the background colour and the colour of the dropdown glyph:

image image

Here're the captured records: [enabled.txt](https://github.com/dotnet/winforms/files/5714874/enabled.txt) [disabled.txt](https://github.com/dotnet/winforms/files/5714875/disabled.txt)

image


Are we missing any records? Or is there something I'm not seeing?

**Comment by RussKie** [comment]
Furthermore, with visual styles on:
image image

Both of these produce the same set of records:
```
[ENHMETAHEADER] Bounds: {0, 0, 199, 22 (LTRB)} Device Size: {Width=3840, Height=2160} Header Size: 108
[EMRINTERSECTCLIPRECT] RECT: {0, 0, 200, 23 (LTRB)}
[EMRBITBLT] Bounds: {0, 0, 199, 22 (LTRB)} Destination: {0, 0, 200, 23 (LTRB)} Rop: SRCCOPY Source DC Color: [R=0, G=120, B=215] (COLOR_HIGHLIGHT)
[EMREXTSELECTCLIPRGN] Mode: COPY Bounds: {0, 0, 200, 23 (LTRB)} Rects: 1
	Rect index 0: {0, 0, 200, 23 (LTRB)}
[EMREOF]
```

Is there any magic we can sprinkle to test? Or should we just forgo these tests for now?

**Comment by tbolon** [comment]
> Is there any magic we can sprinkle to test? Or should we just forgo these tests for now?
Maybe we could just render the control using `DrawToBitmap` and do some pixel color tests? I can write a small helper class.

**Comment by RussKie** [comment]
> Maybe we could just render the control using `DrawToBitmap` and do some pixel color tests? I can write a small helper class.

In the past we had discussed ideas of rendering to bitmaps and assert against those using something like Verify or ApprovalTests, but none of those seemed sufficiently flexible and scalable, e.g. controls may look differently depending on OS version, style/theme/dpi settings, etc...

We haven't certainly closed the door on these discussions, and welcome fresh ideas.

**Comment by JeremyKuhne** [comment]
> Are we missing any records? Or is there something I'm not seeing?

You're only capturing the text output. We're not doing the painting here, you'd have to debug the native control to see what it is doing with the WM_PRINT and WM_PAINT messages.

Because we aren't doing drawing ourselves you'd probably be better off checking the properties rather than looking at what exactly it is drawing under the covers, which doesn't help you with styles anyway. More fundamentally, the problem here is that the native control caches the window style and uses that to determine whether or not to draw the border disabled when theming is enabled. There isn't any state that I can see exposed that you can determine you're in a stale state.

While recreating the control updates the cached state, it is a little heavy handed. You can get it to update it's cached style info by just setting the window style to itself. This seems to work fine to me:

``` C#
        protected override void OnEnabledChanged(EventArgs e)
        {
            base.OnEnabledChanged(e);

            if (Application.RenderWithVisualStyles)
            {
                // The SysDateTimePick32 control caches the style and uses that directly to determine whether the
                // border should be drawn disabled when theming (VisualStyles) is enabled. Setting the window
                // style to itself (which will have the proper WS_DISABLED setting after calling base) will
                // flush the cached value and render the border as one would expect.
                User32.SetWindowLong(this, User32.GWL.STYLE, User32.GetWindowLong(this, User32.GWL.STYLE));
            }
        }
```

I'd consider doing that and just chalk up automated testing for it as impractical. Maybe in the future we can find a reasonable way to incorporate something like https://github.com/microsoft/Detours so that we can look at the UxTheme calls the control is making to validate fixes like this.

**Comment by tbolon** [comment]
Sorry for having left this code in limbo...

I agree that recreating a handle could have been really costly to just update the border style... I was not really satisfied with this solution.

I have updated my PR:
- by removing the automated test, as the non-visualstyle rendering is the same disabled or enabled and visual styles are not testable
- by using the suggested code from @JeremyKuhne

**Review by JeremyKuhne — APPROVED** [review]
Looks good to me. Thanks for validating that it works correctly!

**Comment by RussKie** [comment]
Thank you!

File: src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs [hunk 5]
```diff
@@ -1195,6 +1195,21 @@ protected override void OnHandleDestroyed(EventArgs e)
...
+        {
+            base.OnEnabledChanged(e);
+
+            if (Application.RenderWithVisualStyles)
+            {
+                // The SysDateTimePick32 control caches the style and uses that directly to determine whether the
+                // border should be drawn disabled when theming (VisualStyles) is enabled. Setting the window
+                // style to itself (which will have the proper WS_DISABLED setting after calling base) will
+                // flush the cached value and render the border as one would expect.
+                User32.SetWindowLong(this, User32.GWL.STYLE, User32.GetWindowLong(this, User32.GWL.STYLE));
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1209]: Won't this force the creation of the control's handle?

Comment (tbolon) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1209]: If you are right, then this should be handled (by checking IsHandleCreated) in the condition.
Creating a handle just by setting Enabled to false is certainly not desired.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1209]: 🤦‍♂️

Comment (tbolon) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1209]: Should I create another issue and create a PR (after confirming the behavior when the handle is not created: we should just create an instance of the control, subscribe to HandleCreated event, set Enabled = false and check if the event is triggered or not) ? Or someone with a direct access to the repo could fix this directly ?

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1209]: You can just raise a PR referencing this thread.

Comment (tbolon) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:1209]: I can confirm that the problem is present:

image

I will submit a PR.

**Pull request merged**